host/switchless_manager.c

```diff
@@ -0,0 +1,172 @@
+/* Copyright (c) Microsoft Corporation.  All rights reserved.
+ * Licensed under the MIT License. */
+
+#include "switchless_manager.h"
+#include <openenclave/internal/calls.h>
+#include "sgx/enclave.h"
+
+/* thread helper functions */
+#if _MSC_VER
+
+#define _thread_return_t DWORD WINAPI
+typedef LPVOID _thread_arg_type;
+#define _THREAD_RETURN_VAL 0
+
+typedef DWORD (*_thread_op_t)(LPVOID);
+
+static int _thread_create(
+    oe_thread_t* thread,
+    _thread_op_t op,
+    _thread_arg_t arg)
+{
+    *thread = CreateThread(NULL, 0, op, arg, 0, NULL);
+    return NULL == *thread;
+} /* _thread_create */
+
+static void _thread_join(_thread_arg_t thread)
+{
+    WaitForSingleObject(thread, INFINITE);
+    CloseHandle(thread);
+} /* _thread_join */
+
+#elif defined __GNUC__
+
+typedef void* _thread_return_t;
+typedef void* _thread_arg_t;
+#define _THREAD_RETURN_VAL NULL
+
+typedef void* (*_thread_op_t)(void*);
+
+static int _thread_create(
+    oe_thread_t* thread,
+    _thread_op_t op,
+    _thread_arg_t arg)
+{
+    return pthread_create(thread, NULL, op, arg);
+} /* _thread_create */
+
+static void _thread_join(oe_thread_t thread)
+{
+    pthread_join(thread, NULL);
+} /* _thread_join */
+
+#endif /* _MSC_VER or __GNUC__ */
+
+static _thread_return_t _enclave_worker_thread(_thread_arg_t arg)
+{
+    oe_enclave_t* enclave = (oe_enclave_t*)arg;
+
+    oe_ecall(
+        enclave,
+        OE_ECALL_LAUNCH_ENCLAVE_WORKER,
+        (uint64_t)(&enclave->switchless_manager.switchless),
+        NULL);
+
+    return _THREAD_RETURN_VAL;
+} /* _enclave_worker_thread */
+
+void oe_switchless_manager_init(oe_enclave_t* enclave)
+{
+    enclave->switchless_manager.switchless.state = OE_SWITCHLESS_STATE_STOPPED;
+    oe_lockless_queue_init(
+        &(enclave->switchless_manager.switchless.ecall_queue));
+} /* oe_switchless_manager_init */
+
+oe_result_t oe_switchless_manager_startup(oe_enclave_t* enclave)
+{
+    oe_result_t result = OE_OK;
+    oe_switchless_state state = OE_SWITCHLESS_STATE_STOPPED;
+
+    /* check the current state:
+       - if it is stopped then change to starting */
+#ifdef _MSC_VER
+    state = (oe_switchless_state)_InterlockedCompareExchange(
+        &(enclave->switchless_manager.switchless.state),
+        OE_SWITCHLESS_STATE_STARTING,
+        OE_SWITCHLESS_STATE_STOPPED);
+#elif defined __GNUC__
+    __atomic_compare_exchange_n(
+        &(enclave->switchless_manager.switchless.state),
+        &state,
+        OE_SWITCHLESS_STATE_STARTING,
+        1,
+        __ATOMIC_ACQ_REL,
```

Comment:
Can you describe why you prefer __ATOMIC_ACQ_REL instead of __ATOMIC_RELEASE for the success case?
We can aim for consistent use of acquire/release everywhere.

Reply:
__atomic_compare_exchange_n is read-then-write operation.  In both the success and failure cases an __ATOMIC_ACQUIRE is performed to refresh the current value.  The success case writes to the value necessitating an __ATOMIC_RELEASE as well.  The failure case does not modify the value.  So the success case needs __ATOMIC_ACQ_REL where the failure case only needs __ATOMIC_ACQUIRE.

Reply:
> __ATOMIC_ACQUIRE is performed to refresh the current value
I don't quite follow. The current value of the variable being manipulated will always be refreshed due to the atomic operation, isn't it?

Couple of suggestions:
- It would be better to capture these rationales in code comments as opposed to PR comments so that a future reader can easily refer to them in source file without having to search for this PR in github.
- Is oe_switchless_manager_startup expected to be called only once? If so, it is better to wrap the call in oe_once

Reply:
- I've changed the operations to add more comments
- I've changed oe_switchless_manager_startup/shutdown to be re-entrant.  Neither function will return until the switchless manager is in the desired state (with the exception of thread creation failure).
- I've added comments to explain the acquire/release semantics.
- oe_switchless_manager_startup now returns OE_OK in all cases unless thread create fails.

- oe_once is not needed as re-entrant calls are handled.  Additionally, oe_once would not work with multiple enclaves, or if the enclave is terminated and then created.